base/regex.jl

```diff
@@ -422,11 +422,11 @@ struct SubstitutionString{T<:AbstractString} <: AbstractString
     string::T
 end
 
-ncodeunits(s::SubstitutionString) = ncodeunits(s.string)
-codeunit(s::SubstitutionString) = codeunit(s.string)
-codeunit(s::SubstitutionString, i::Integer) = codeunit(s.string, i)
-isvalid(s::SubstitutionString, i::Integer) = isvalid(s.string, i)
-iterate(s::SubstitutionString, i::Integer...) = iterate(s.string, i...)
+ncodeunits(s::SubstitutionString) = ncodeunits(s.string)::Int
+codeunit(s::SubstitutionString) = codeunit(s.string)::Type{<:Union{UInt8, UInt16, UInt32}}
+codeunit(s::SubstitutionString, i::Integer) = codeunit(s.string, i)::Union{UInt8, UInt16, UInt32}
```

Comment:
I'm pretty sure this is true for all string types in practice, but it's not technically required.

Reply:
Docs say this:
```julia
codeunit(s::AbstractString) -> Type{<:Union{UInt8, UInt16, UInt32}}
```
I've been taking the docs as license to enforce stuff. Should I not?

Reply:
I guess someone could have a different code unit, but it would be very weird. Not sure all of the logic in base would work correctly with some other code unit.